auth/src/main/java/com/firebase/ui/auth/ui/email/PreambleHandler.java

```diff
@@ -0,0 +1,122 @@
+package com.firebase.ui.auth.ui.email;
+
+import android.content.Context;
+import android.net.Uri;
+import android.support.annotation.ColorInt;
+import android.support.annotation.StringRes;
+import android.support.customtabs.CustomTabsIntent;
+import android.support.v4.content.ContextCompat;
+import android.text.SpannableStringBuilder;
+import android.text.TextUtils;
+import android.text.method.LinkMovementMethod;
+import android.text.style.ClickableSpan;
+import android.text.style.ForegroundColorSpan;
+import android.util.TypedValue;
+import android.view.View;
+import android.widget.TextView;
+
+import com.firebase.ui.auth.R;
+import com.firebase.ui.auth.ui.FlowParameters;
+
+public class PreambleHandler {
+    private static final String TOS_TARGET = "%TOS%";
+    private static final String PP_TARGET = "%PP%";
+
+    private final Context mContext;
+    private final FlowParameters mFlowParameters;
+    private final ForegroundColorSpan mLinkSpan;
+
+    private SpannableStringBuilder mBuilder;
+
+    public PreambleHandler(Context context, FlowParameters parameters) {
+        mContext = context;
+        mFlowParameters = parameters;
+        mLinkSpan = new ForegroundColorSpan(ContextCompat.getColor(mContext, R.color.linkColor));
+
+        setupCreateAccountPreamble();
+    }
+
+    public void setPreamble(TextView textView) {
+        textView.setMovementMethod(LinkMovementMethod.getInstance());
+        textView.setText(mBuilder);
+    }
+
+    private void setupCreateAccountPreamble() {
+        int preambleType = getPreambleType();
+        if (preambleType == -1) {
+            return;
+        }
+
+        String[] preambles =
+                mContext.getResources().getStringArray(R.array.create_account_preamble);
+        mBuilder = new SpannableStringBuilder(preambles[preambleType]);
+
+        replaceTarget(TOS_TARGET, R.string.terms_of_service, mFlowParameters.termsOfServiceUrl);
+        replaceTarget(PP_TARGET, R.string.privacy_policy, mFlowParameters.privacyPolicyUrl);
+    }
+
+    private void replaceTarget(String target, @StringRes int replacementRes, String url) {
+        char[] currentPreambleChars = new char[mBuilder.length()];
+        mBuilder.getChars(0, mBuilder.length(), currentPreambleChars, 0);
+        String currentPreamble = String.valueOf(currentPreambleChars);
+
+        int targetIndex = currentPreamble.indexOf(target);
```

Comment:
@TheCraftKid whoops, forgot to mention that I updated my patch. Turns out the `SpannableStringBuilder#toString()` method does exactly what I just did so all of that can be replaced with `mBuilder.toString().indexOf(target);`